### PR TITLE
Add setIndexConfiguration API

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
Ports https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java#L307

Index API is not exported to the main `index.ts` file and is hence still invisible.

### CODE REVIEW

File: packages/firestore/src/api/database.ts [hunk 1]
```diff
@@ -503,13 +503,13 @@ export function terminate(firestore: Firestore): Promise<void> {
 /**
  * Loads a Firestore bundle into the local cache.
  *
- * @param firestore - The {@link Firestore} instance to load bundles for for.
- * @param bundleData - An object representing the bundle to be loaded. Valid objects are
- *   `ArrayBuffer`, `ReadableStream<Uint8Array>` or `string`.
+ * @param firestore - The {@link Firestore} instance to load bundles for.
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts:506]: Wrapped to line length

**Review by thebrianchen — APPROVED** [review]
lgtm with some suggestions

File: packages/firestore/src/api/index_configuration.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      const segments: Segment[] = [];
+      if (Array.isArray(index.fields)) {
+        for (const field of index.fields) {
+          const fieldPathString = tryGetString(field, 'fieldPath');
+          const fieldPath = fieldPathFromDotSeparatedString(
+            'setIndexConfiguration',
+            fieldPathString
+          );
+
+          if (field.arrayConfig === 'CONTAINS') {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/api/index_configuration.ts:159]: I feel there are lots of ways we can validate here, and it is a bit of a rabbit hole that I don't want to go down just yet. If there is demand, then we can consider.

File: packages/firestore/src/model/field_index.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  constructor(
+    /**
+     * The index ID. Returns -1 if the index ID is not available (e.g. the index
+     * has not yet been persisted).
+     */
+    readonly indexId: number,
+    /** The collection ID this index applies to. */
+    readonly collectionGroup: string,
+    /** Returns all field segments for this index. */
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_index.ts]: Done

**Pull request merged**